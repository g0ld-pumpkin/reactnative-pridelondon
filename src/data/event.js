// @flow
import R from "ramda";
import {
  set as setDate,
  diff as diffDate,
  add as addToDate,
  toFormat as formatDate,
  FORMAT_EUROPEAN_DATE,
  FORMAT_CONTENTFUL_ISO
} from "../lib/date";
import type { Maybe } from "../lib/maybe";
import * as maybe from "../lib/maybe";
import type { Decoder } from "../lib/decode";
import * as decode from "../lib/decode";
import type { FieldRef } from "./field-ref";
import decodeFieldRef from "./field-ref";
import * as colors from "../constants/colors";

export type Events = {
  [id: string]: Event
};

export type EventCategoryName =
  | "Cabaret and Variety"
  | "Community"
  | "Talks and Debates"
  | "Film and Screenings"
  | "Plays and Theatre"
  | "Social and Networking"
  | "Nightlife"
  | "Exhibition and Tours"
  | "Sports and Activities"
  | "Health"
  | "Music";

export const eventCategoryNames = [
  "Cabaret and Variety",
  "Community",
  "Talks and Debates",
  "Film and Screenings",
  "Plays and Theatre",
  "Social and Networking",
  "Nightlife",
  "Exhibition and Tours",
  "Sports and Activities",
  "Health",
  "Music"
];

const decodeEventCategoryName = decode.oneOf(
  eventCategoryNames.map(decode.value)
);

export type EventCategory = {
  label: EventCategoryName,
  color: string,
  contrast: boolean
};

// We use a switch here, rather than a look up so that we can ensure that
// this function will always return an EventCategory.
export const getEventCategoryFromName = (
  category: EventCategoryName
): EventCategory => {
  switch (category) {
    case "Cabaret and Variety":
      return {
        label: "Cabaret and Variety",
        color: colors.coralColor,
        contrast: false
      };
    case "Community":
      return {
        label: "Community",
        color: colors.brightLightBlueColor,
        contrast: true
      };
    case "Talks and Debates":
      return {
        label: "Talks and Debates",
        color: colors.darkSkyBlueColor,
        contrast: false
      };
    case "Film and Screenings":
      return {
        label: "Film and Screenings",
        color: colors.lightTealColor,
        contrast: true
      };
    case "Plays and Theatre":
      return {
        label: "Plays and Theatre",
        color: colors.warmPinkColor,
        contrast: false
      };
    case "Social and Networking":
      return {
        label: "Social and Networking",
        color: colors.turquoiseBlueColor,
        contrast: true
      };
    case "Nightlife":
      return {
        label: "Nightlife",
        color: colors.yellowColor,
        contrast: true
      };
    case "Exhibition and Tours":
      return {
        label: "Exhibition and Tours",
        color: colors.brightPurpleColor,
        contrast: false
      };
    case "Sports and Activities":
      return {
        label: "Sports and Activities",
        color: colors.vomitYellowColor,
        contrast: true
      };
    case "Health":
      return {
        label: "Health",
        color: colors.bubblegumPinkColor,
        contrast: true
      };
    case "Music":
      return {
        label: "Music",
        color: colors.cornflowerBlueColor,
        contrast: false
      };
    default:
      // Default case with cast to enable flow exhaustive match. This will
      // cause flow to error if passed anything other than an EventCategory;
      return (category: empty);
  }
};

export type Event = {
  // important to keep this at the top level so type refinement works
  contentType: "event",
  id: string,
  locale: string,
  revision: number,
  fields: {
    name: string,
    eventCategories: Array<EventCategoryName>,
    audience: Array<string>,
    startTime: string,
    endTime: string,
    location: { lat: number, lon: number },
    addressLine1: Maybe<string>,
    addressLine2: Maybe<string>,
    city: Maybe<string>,
    postcode: Maybe<string>,
    locationName: string,
    eventPriceLow: number,
    eventPriceHigh: number,
    accessibilityOptions: Array<string>,
    eventDescription: string,
    accessibilityDetails: Maybe<string>,
    email: Maybe<string>,
    phone: Maybe<string>,
    ticketingUrl: Maybe<string>,
    venueDetails: Array<string>,
    individualEventPicture: FieldRef,
    eventsListPicture: FieldRef,
    performances: Array<FieldRef>,
    recurrenceDates: Array<string>,
    stage: boolean
  }
};

export type SavedEvents = Set<string>;

export type EventDays = Event[][];

const maybeField = <A>(
  locale: string,
  field: string,
  decoder: Decoder<A>
): Decoder<Maybe<A>> =>
  decode.field(field, decode.maybe(decode.field(locale, decoder)));

const maybeFieldWithDefault = <A>(
  locale: string,
  field: string,
  decoder: Decoder<A>,
  defaultValue: A
): Decoder<A> =>
  decode.map(
    maybe.withDefault(defaultValue),
    maybeField(locale, field, decoder)
  );

export const decodeEvent = (locale: string): Decoder<Event> =>
  decode.shape({
    contentType: decode.at(
      ["sys", "contentType", "sys", "id"],
      decode.value("event")
    ),
    id: decode.at(["sys", "id"], decode.string),
    locale: decode.succeed(locale),
    revision: decode.at(["sys", "revision"], decode.number),
    fields: decode.field(
      "fields",
      decode.shape({
        name: decode.at(["name", locale], decode.string),
        eventCategories: decode.at(
          ["eventCategories", locale],
          decode.array(decodeEventCategoryName)
        ),
        audience: maybeFieldWithDefault(
          locale,
          "audience",
          decode.array(decode.string),
          []
        ),
        // may want to combine startTime and endTime to fix the
        // date flipping issue in the decoder
        // may also want to convert it to a DateTime
        startTime: decode.at(["startTime", locale], decode.string),
        endTime: decode.at(["endTime", locale], decode.string),
        location: decode.at(
          ["location", locale],
          decode.shape({
            lat: decode.field("lat", decode.number),
            lon: decode.field("lon", decode.number)
          })
        ),
        addressLine1: maybeField(locale, "addressLine1", decode.string),
        addressLine2: maybeField(locale, "addressLine2", decode.string),
        city: maybeField(locale, "city", decode.string),
        postcode: maybeField(locale, "postcode", decode.string),
        locationName: decode.at(["locationName", locale], decode.string),
        eventPriceLow: decode.at(["eventPriceLow", locale], decode.number),
        eventPriceHigh: decode.at(["eventPriceHigh", locale], decode.number),
        accessibilityOptions: maybeFieldWithDefault(
          locale,
          "accessibilityOptions",
          decode.array(decode.string),
          []
        ),
        eventDescription: decode.at(
          ["eventDescription", locale],
          decode.string
        ),
        accessibilityDetails: maybeField(
          locale,
          "accessibilityDetails",
          decode.string
        ),
        email: maybeField(locale, "email", decode.string),
        phone: maybeField(locale, "phone", decode.string),
        ticketingUrl: maybeField(locale, "ticketingUrl", decode.string),
        venueDetails: maybeFieldWithDefault(
          locale,
          "venueDetails",
          decode.array(decode.string),
          []
        ),
        individualEventPicture: decode.at(
          ["individualEventPicture", locale],
          decodeFieldRef
        ),
        eventsListPicture: decode.at(
          ["eventsListPicture", locale],
          decodeFieldRef
        ),
        performances: maybeFieldWithDefault(
          locale,
          "performances",
          decode.array(decodeFieldRef),
          []
        ),
        recurrenceDates: maybeFieldWithDefault(
          locale,
          "recurrenceDates",
          decode.array(decode.string),
          []
        ),
        stage: maybeFieldWithDefault(locale, "stage", decode.boolean, false)
      })
    )
  });

const formatEuropeanDate = value => formatDate(value, FORMAT_EUROPEAN_DATE);

const generateRecurringEvent = (event: Event) => (
  recurrenceStartTime: string
) => {
  const { endTime, startTime } = event.fields;

  const difference = diffDate(recurrenceStartTime, startTime);
  const recurrenceEndTime = addToDate(endTime, difference);

  return R.mergeDeepRight(event, {
    fields: {
      startTime: formatDate(recurrenceStartTime, FORMAT_CONTENTFUL_ISO),
      endTime: formatDate(recurrenceEndTime, FORMAT_CONTENTFUL_ISO),
      recurrenceDates: [
        formatEuropeanDate(startTime),
        ...event.fields.recurrenceDates
      ]
    },
    id: `${event.id}-recurrence-${formatEuropeanDate(recurrenceStartTime)}`
  });
};

const recurrenceDateToStartTime = (originalStartTime: string) => (
  recurrence: string
) => {
  const [recurrenceDay, recurrenceMonth, recurrenceYear] = recurrence.split(
    "/"
  );

  return formatDate(
    setDate(originalStartTime, {
      year: recurrenceYear,
      month: recurrenceMonth,
      day: recurrenceDay
    }),
    FORMAT_CONTENTFUL_ISO
  );
};

export const expandRecurringEvents = (event: Event): Array<Event> => {
  const recurrenceStartTimes = [
    event.fields.startTime,
    ...event.fields.recurrenceDates.map(
      recurrenceDateToStartTime(event.fields.startTime)
    )
  ];

  const events = R.uniq(recurrenceStartTimes)
    .slice(1)
    .map(generateRecurringEvent(event));
  return [event, ...events];
};
