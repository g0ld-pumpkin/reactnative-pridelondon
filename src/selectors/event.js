// @flow
import R from "ramda";
import { compareAsc as compareDateAsc, isSameDay } from "../lib/date";
import type { Event, EventDays } from "../data/event";

export const isFree = (priceLow: number, priceHigh: number) =>
  priceLow === 0 && priceHigh === 0;

export const selectEventIsFree = (event: Event) =>
  isFree(event.fields.eventPriceLow, event.fields.eventPriceHigh);

const sortByStartTimeAsc = (a: Event, b: Event) =>
  compareDateAsc(a.fields.startTime, b.fields.startTime);

export const groupEventsByStartTime = (events: Event[]): EventDays =>
  R.groupWith(
    (a: Event, b: Event) => isSameDay(a.fields.startTime, b.fields.startTime),
    events.sort(sortByStartTimeAsc)
  );

export const filterEvents = (
  events: Event[],
  filter: Event => boolean
): Event[] => events.filter(filter);

export const getStages = (events: Event[]) =>
  events.filter(event => event.fields.stage);
