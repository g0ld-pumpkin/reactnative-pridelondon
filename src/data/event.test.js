// @flow
import {
  generateCMSEvent,
  generateCMSEventMinimum,
  generateEvent,
  sampleOne
} from "./__test-data";
import { decodeEvent, expandRecurringEvents } from "./event";

describe("Event", () => {
  describe("decoder", () => {
    it("correctly decodes valid CMS event with all fields", () => {
      const data: mixed = sampleOne(generateCMSEvent);

      const decoded = decodeEvent("en-GB")(data);
      expect(decoded.ok).toEqual(true);
      if (decoded.ok) {
        expect(decoded.value).toMatchSnapshot();
      }
    });

    it("correctly decodes valid CMS event with minimum fields", () => {
      const data: mixed = sampleOne(generateCMSEventMinimum);

      const decoded = decodeEvent("en-GB")(data);
      expect(decoded.ok).toEqual(true);
      if (decoded.ok) {
        expect(decoded.value).toMatchSnapshot();
      }
    });

    it("fails if a property is missing", () => {
      const data: mixed = {
        fields: {},
        sys: {
          id: "3O3SZPgYl2MUEWu2MoK2oi",
          contentType: {
            sys: {
              id: "event"
            }
          }
        }
      };

      const decoded = decodeEvent("en-GB")(data);
      expect(decoded.ok).toEqual(false);
    });
  });

  describe("expandRecurringEvents", () => {
    it("creates new events for each recurrence date", () => {
      const event = sampleOne(generateEvent);
      event.id = "test";
      event.fields.startTime = "2018-08-02T00:00+00:00";
      event.fields.endTime = "2018-08-02T03:00+00:00";
      event.fields.recurrenceDates = ["03/08/2018", "04/08/2018"];

      const expandedEvents = expandRecurringEvents(event);

      expect(expandedEvents.length).toEqual(3);
      expect(expandedEvents[0]).toEqual(event);
      expect(expandedEvents[1].id).toEqual("test-recurrence-03/08/2018");
      expect(expandedEvents[1].fields.startTime).toEqual(
        "2018-08-03T00:00+00:00"
      );
      expect(expandedEvents[1].fields.recurrenceDates).toEqual([
        "02/08/2018",
        "03/08/2018",
        "04/08/2018"
      ]);
      expect(expandedEvents[2].id).toEqual("test-recurrence-04/08/2018");
      expect(expandedEvents[2].fields.startTime).toEqual(
        "2018-08-04T00:00+00:00"
      );
      expect(expandedEvents[2].fields.recurrenceDates).toEqual([
        "02/08/2018",
        "03/08/2018",
        "04/08/2018"
      ]);
    });

    it("creates new events for short format recurrence dates", () => {
      const event = sampleOne(generateEvent);
      event.id = "test";
      event.fields.startTime = "2018-08-02T00:00+00:00";
      event.fields.endTime = "2018-08-02T03:00+00:00";
      event.fields.recurrenceDates = ["3/8/2018", "4/8/2018"];

      const expandedEvents = expandRecurringEvents(event);

      expect(expandedEvents.length).toEqual(3);
      expect(expandedEvents[0]).toEqual(event);
      expect(expandedEvents[1].id).toEqual("test-recurrence-03/08/2018");
      expect(expandedEvents[1].fields.startTime).toEqual(
        "2018-08-03T00:00+00:00"
      );
      expect(expandedEvents[1].fields.recurrenceDates).toEqual([
        "02/08/2018",
        "3/8/2018",
        "4/8/2018"
      ]);
      expect(expandedEvents[2].id).toEqual("test-recurrence-04/08/2018");
      expect(expandedEvents[2].fields.startTime).toEqual(
        "2018-08-04T00:00+00:00"
      );
      expect(expandedEvents[1].fields.recurrenceDates).toEqual([
        "02/08/2018",
        "3/8/2018",
        "4/8/2018"
      ]);
    });

    it("updates endTime to be same distance from startTime", () => {
      const event = sampleOne(generateEvent);
      event.id = "test";
      event.fields.startTime = "2018-08-02T00:00+00:00";
      event.fields.endTime = "2018-08-02T03:00+00:00";
      event.fields.recurrenceDates = ["03/08/2018"];

      const expandedEvents = expandRecurringEvents(event);

      expect(expandedEvents[1].fields.endTime).toEqual(
        "2018-08-03T03:00+00:00"
      );
    });

    it("preserves timezone of startTime and endTime", () => {
      const event = sampleOne(generateEvent);
      event.fields.startTime = "2018-04-19T23:00+14:00";
      event.fields.endTime = "2018-04-20T14:48+02:00";
      event.fields.recurrenceDates = ["25/04/2018"];

      const expandedEvents = expandRecurringEvents(event);

      expect(expandedEvents[1].fields.startTime).toEqual(
        "2018-04-25T23:00+14:00"
      );
      expect(expandedEvents[1].fields.endTime).toEqual(
        "2018-04-26T14:48+02:00"
      );
    });

    it("does not create new events recurrence is on the same day as start date", () => {
      const event = sampleOne(generateEvent);
      event.id = "test";
      event.fields.startTime = "2018-08-02T00:00+00:00";
      event.fields.endTime = "2018-08-02T03:00+00:00";
      event.fields.recurrenceDates = [
        "02/08/2018",
        "2/8/2018",
        "2/08/2018",
        "02/8/2018"
      ];

      const expandedEvents = expandRecurringEvents(event);

      expect(expandedEvents.length).toEqual(1);
      expect(expandedEvents[0]).toEqual(event);
    });
  });
});
