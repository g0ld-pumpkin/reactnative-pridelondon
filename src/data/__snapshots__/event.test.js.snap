// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event decoder correctly decodes valid CMS event with all fields 1`] = `
Object {
  "contentType": "event",
  "fields": Object {
    "accessibilityDetails": "accessibilityDetails",
    "accessibilityOptions": Array [
      "WG3dQjnkVo",
      "XLA8X82wzzA6",
      "BHI9hy12J10cm",
      "8zA9aus40c6Ael46C99709HocLl",
    ],
    "addressLine1": "addressLine1",
    "addressLine2": "addressLine2",
    "audience": Array [
      "Vglv35495N98k",
      "2",
      "rDY8qmFO1oz17mbZKz19fHgB0Hn",
      "be82idS3Rk5O7mBq1X",
    ],
    "city": "city",
    "email": "email",
    "endTime": "2018-07-07T03:00+00:00",
    "eventCategories": Array [
      "Film and Screenings",
      "Exhibition and Tours",
      "Sports and Activities",
    ],
    "eventDescription": "eventDescription",
    "eventPriceHigh": 10,
    "eventPriceLow": 0,
    "eventsListPicture": Object {
      "sys": Object {
        "id": "SOibJoT",
      },
    },
    "individualEventPicture": Object {
      "sys": Object {
        "id": "8qQ6Xn8vg3g",
      },
    },
    "location": Object {
      "lat": 0,
      "lon": 10,
    },
    "locationName": "locationName",
    "name": "name",
    "performances": Array [
      Object {
        "sys": Object {
          "id": "430A34PrvRPgzjlX98",
        },
      },
      Object {
        "sys": Object {
          "id": "My2ocdeOtW",
        },
      },
    ],
    "phone": "phone",
    "postcode": "postcode",
    "recurrenceDates": Array [
      "09/07/2018",
      "07/07/2018",
      "09/06/2018",
      "24/07/2018",
      "08/08/2018",
    ],
    "stage": false,
    "startTime": "2018-07-07T00:00+00:00",
    "ticketingUrl": "ticketingUrl",
    "venueDetails": Array [
      "6LbKV5M",
      "GdB991VXq0FXL",
      "",
      "91kU05Y047r",
    ],
  },
  "id": "hG",
  "locale": "en-GB",
  "revision": 1,
}
`;

exports[`Event decoder correctly decodes valid CMS event with minimum fields 1`] = `
Object {
  "contentType": "event",
  "fields": Object {
    "accessibilityDetails": null,
    "accessibilityOptions": Array [],
    "addressLine1": null,
    "addressLine2": null,
    "audience": Array [],
    "city": null,
    "email": null,
    "endTime": "2018-07-07T03:00+00:00",
    "eventCategories": Array [
      "Cabaret and Variety",
      "Music",
    ],
    "eventDescription": "eventDescription",
    "eventPriceHigh": 10,
    "eventPriceLow": 0,
    "eventsListPicture": Object {
      "sys": Object {
        "id": "9NOp294Sf3T7655",
      },
    },
    "individualEventPicture": Object {
      "sys": Object {
        "id": "QPIB7DYkm7Z",
      },
    },
    "location": Object {
      "lat": 0,
      "lon": 10,
    },
    "locationName": "locationName",
    "name": "name",
    "performances": Array [],
    "phone": null,
    "postcode": null,
    "recurrenceDates": Array [],
    "stage": false,
    "startTime": "2018-07-07T00:00+00:00",
    "ticketingUrl": null,
    "venueDetails": Array [],
  },
  "id": "hG",
  "locale": "en-GB",
  "revision": 1,
}
`;
